src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,416 @@
+            targetNodeType = targetNodeInfo.Type;
+
+            if (targetNodeType == null)
+                return false;
+
+            // The error happens either on an assignement operation or on an invocation expression.
+            // If the error happens on assignment operation, "ConvertedType" is different from the current "Type"
+            using var _ = ArrayBuilder<ITypeSymbol>.GetInstance(out var mutablePotentialConversionTypes);
+            if (diagnosticId == CS0266
+                && targetNodeInfo.ConvertedType != null
+                && !targetNodeType.Equals(targetNodeInfo.ConvertedType))
+            {
+                mutablePotentialConversionTypes.Add(targetNodeInfo.ConvertedType);
+            }
+            else if (diagnosticId == CS1503
+                && targetNode.GetAncestorsOrThis<ArgumentSyntax>().FirstOrDefault() is ArgumentSyntax targetArgument
+                && targetArgument.Parent is ArgumentListSyntax argumentList
+                && argumentList.Parent is SyntaxNode invocationNode) // invocation node could be Invocation Expression, Object Creation, Base Constructor...
+            {
+                // Implicit downcast appears on the argument of invocation node, get all candidate functions and extract potential conversion types 
+                var symbolInfo = semanticModel.GetSymbolInfo(invocationNode, cancellationToken);
+                var candidateSymbols = symbolInfo.CandidateSymbols;
+
+                foreach (var candidateSymbol in candidateSymbols.OfType<IMethodSymbol>())
+                {
+                    if (CanArgumentTypesBeConvertedToParameterTypes(semanticModel, argumentList.Arguments,
+                        candidateSymbol.Parameters, targetArgument, cancellationToken, out var paramIndex))
+                    {
+                        var correspondingParameter = candidateSymbol.Parameters[paramIndex];
+                        var argumentConversionType = correspondingParameter.Type;
+
+                        if (correspondingParameter.IsParams
+                            && correspondingParameter.Type is IArrayTypeSymbol arrayType
+                            && !(targetNodeType is IArrayTypeSymbol))
+                        {
+                            // target argument is matched to the parameter with keyword params
+                            argumentConversionType = arrayType.ElementType;
+                        }
+
+                        mutablePotentialConversionTypes.Add(argumentConversionType);
+                    }
+                }
+
+                // Sort the potential conversion types by inheritance distance, so that
+                // operations are in order and user can choose least specific types(more accurate)
+                var comparer = new InheritanceDistanceComparer(semanticModel, targetNodeType);
+                mutablePotentialConversionTypes.Sort(comparer);
+            }
```

Comment:
consider extracting this entirely into a seprate helper method to keep this top level method much clearer.

Reply:
you resolved this comment without a change or comment of your own :)

Reply:
Oh, I move 2 lines of code into a helper method below, do I also need to move the comments?

Reply:
ack.  i've given you conflicting inf :D

Reply:
So here's what i was trying to say:

Extract the *whole* block into a helper method, not the two statements above this.